File: neet_code_all_cpp/math_and_geometry/excel_sheet_column_title.cpp
#include <iostream>
#include <string>
#include <algorithm>

class SolutionAnsPython {
    public:
        std::string convertToTitle(int columnNumber) {
            auto result = std::string{};
            while(columnNumber > 0) {
                auto tmp = (columnNumber - 1) % 26;
                result.push_back('A' + tmp);
                columnNumber = (columnNumber - 1) / 26;
            }

            std::reverse(result.begin(), result.end());
            return result;
        }
};

int main(void) {
    int case_1 = 1;
    // "A"
    int case_2 = 28;
    // "AB"
    int case_3 = 701;
    // "ZY"

    SolutionAnsPython s_ans_py;

    std::cout << s_ans_py.convertToTitle(case_1) << std::endl;
    std::cout << s_ans_py.convertToTitle(case_2) << std::endl;
    std::cout << s_ans_py.convertToTitle(case_3) << std::endl;
}

File: neet_code_all/src/math_and_geometry/excel_sheet_column_title.rs
struct Solution;
impl Solution {
    pub fn convert_to_title(column_number: i32) -> String {
        let mut column_number = column_number;
        let num_to_alphabet = |n: i32| (b'A' + (n as u8 - 1)) as char;

        let mut result = "".to_string();
        /*
        if column_number <= 26 {
            let c_res = num_to_alphabet(column_number);
            result.push(c_res);
        }
        */
        while column_number > 0 {
            let (q, r) = (column_number / 26, column_number % 26);
            println!("q = {q}, r = {r}");
            column_number /= 26;
        }

        result
    }
}

struct SolutionAnsPython;
impl SolutionAnsPython {
    pub fn convert_to_title(column_number: i32) -> String {
        let mut column_number = column_number;
        let mut result = vec![];
        while column_number > 0 {
            let tmp = ((column_number - 1) % 26) as u8;
            result.push((b'A' + tmp) as char);
            column_number = (column_number - 1) / 26;
        }
        result.reverse();

        result.iter().collect()
    }
}

fn main() {
    let case_1 = 1;
    // => "A"
    let case_2 = 28;
    // => "AB"
    let case_3 = 701;
    // => "ZY"

    //println!("case_1: {}", Solution::convert_to_title(case_1));
    //println!("case_3: {}", Solution::convert_to_title(case_3));

    println!("case_1: {}", SolutionAnsPython::convert_to_title(case_1));
    println!("case_2: {}", SolutionAnsPython::convert_to_title(case_2));
    println!("case_3: {}", SolutionAnsPython::convert_to_title(case_3));
}
